macros/edr/dbt_artifacts/upload_artifacts_to_table.sql

```diff
@@ -18,4 +18,61 @@
     {%- if should_commit -%}
         {% do adapter.commit() %}
     {%- endif -%}
-{% endmacro %}
\ No newline at end of file
+{% endmacro %}
+
+{% macro upload_csv_artifacts_to_table(table_relation, artifacts, flatten_artifact_callback) %}
+    {% set output_csv_path = elementary.get_target_path(table_relation.identifier ~ '.csv') %}
+    {% set flatten_artifact_dicts = [] %}
+    {% for artifact in artifacts %}
+        {% set flatten_artifact_dict = flatten_artifact_callback(artifact) %}
+        {% set flatten_artifact_rendered_dict = {} %}
+        {% for key in flatten_artifact_dict %}
+          {% set value = flatten_artifact_dict[key] %}
+          {% if value %}
+            {% do flatten_artifact_rendered_dict.update({key: elementary.render_csv_value(value)}) %}
+          {% endif %}
+        {% endfor %}
+        {% do flatten_artifact_dicts.append(flatten_artifact_rendered_dict) %}
+    {% endfor %}
+    {% set flatten_artifacts_agate = elementary.get_agate_table().from_object(flatten_artifact_dicts) %}
+    {% set cached_artifacts_agate = elementary.get_cached_artifacts_agate(table_relation, output_csv_path) %}
+    {% if elementary.are_artifacts_equal(cached_artifacts_agate, flatten_artifacts_agate) %}
+      {% do elementary.debug_log("[%s] Artifacts were not changed. Skipping upload." % table_relation.identifier) %}
+      {% do return(none) %}
+    {% endif %}
+    {% do elementary.debug_log("[%s] Uploading artifacts." % table_relation.identifier) %}
+    {% do elementary.seed_elementary_model(table_relation, output_csv_path) %}
+    {% do flatten_artifacts_agate.to_csv(output_csv_path) %}
+{% endmacro %}
+
+{% macro get_cached_artifacts_agate(table_relation, cache_csv_path) %}
+  {% if cache_csv_path.exists() %}
+    {% do elementary.debug_log("[%s] Found artifacts cache file." % table_relation.identifier) %}
+    {% do return(elementary.get_agate_table().from_csv(cache_csv_path)) %}
+  {% endif %}
+  {% set cached_artifacts_agate = dbt.run_query('select * from %s' % table_relation) %}
+  {% set lowercased_cached_artifacts_agate = elementary.lowercase_agate_columns(cached_artifacts_agate) %}
+  {% do lowercased_cached_artifacts_agate.to_csv(output_csv_path) %}
+  {% do return(lowercased_cached_artifacts_agate) %}
+{% endmacro %}
+
+{%- macro render_csv_value(value) -%}
+  {% if not value %}
+    {{ return(none) }}
+  {% endif %}
+
+  {% if value is not string and value is not number %}
+    {{ return(tojson(value)) }}
+  {% else %}
+    {{ return(value) }}
+  {% endif %}
+{% endmacro %}
+
+{% macro are_artifacts_equal(first_artifacts_agate, second_artifacts_agate) %}
+  {% set shared_column_names = set(first_artifacts_agate.column_names).intersection(second_artifacts_agate.column_names) %}
```

Comment:
I think that you should start by comparing the columns, if they are not equal just return False. This will protect us if we add more fields to the artifacts

Reply:
`flatten_test` is currently used for both `dbt_tests` and `elementary_test_results` even though they don't share a column, `compiled_code`, which is bad. This becomes a problem because it attempts to `.select(..., compiled_code)` and then it doesn't find it.

A better solution would be either:

1. In `flatten_test`, check if `compiled_code` exists (it only exists on `RunResult`, not `graph`), only if so add it to the artifact dict.
2. Split `flatten_test` to `flatten_test_artifact` and `flatten_test_result`.

What do you think?

Reply:
Not sure I understand.. you should be comparing the same type of artifact (current vs what's saved in file / DWH) no?